src/basic/fd-util.c

```diff
@@ -421,3 +423,148 @@ int move_fd(int from, int to, int cloexec) {
+         *    earliest boot on). To make it read-only we open the fd a second time with O_RDONLY via
+         *    /proc/self/<fd>. Unfortunately O_TMPFILE is not available on older kernels on tmpfs.
+         *
+         * d) Finally, we try creating a regular file in /dev/shm, which we then delete.
+         *
+         * It sucks a bit that depending on the situation we return very different objects here, but that's Linux I
+         * figure. */
+
+        if (size == 0) {
+                /* As a special case, return /dev/null if we have been called for an empty data block */
+                r = open("/dev/null", O_RDONLY|O_CLOEXEC|O_NOCTTY);
+                if (r < 0)
+                        return -errno;
+
+                return r;
+        }
+
+        fd = memfd_new("data-fd");
+        if (fd < 0)
+                goto try_pipe;
+
+        n = write(fd, data, size);
+        if (n < 0)
+                return -errno;
+        if ((size_t) n != size)
+                return -EIO;
+
+        f = lseek(fd, 0, SEEK_SET);
+        if (f != 0)
+                return -errno;
+
+        r = memfd_set_sealed(fd);
+        if (r < 0)
+                return r;
+
+        r = fd;
+        fd = -1;
+
+        return r;
+
+try_pipe:
+        if (pipe2(pipefds, O_CLOEXEC|O_NONBLOCK) < 0)
+                return -errno;
+
+        isz = fcntl(pipefds[1], F_GETPIPE_SZ, 0);
+        if (isz < 0)
+                return -errno;
+
+        if ((size_t) isz < size) {
+                isz = (int) size;
+                if (isz < 0 || (size_t) isz != size)
+                        return -E2BIG;
+
+                /* Try to bump the pipe size */
+                (void) fcntl(pipefds[1], F_SETPIPE_SZ, isz);
+
+                /* See if that worked */
+                isz = fcntl(pipefds[1], F_GETPIPE_SZ, 0);
+                if (isz < 0)
+                        return -errno;
+
+                if ((size_t) isz < size)
+                        goto try_dev_shm;
+        }
+
+        n = write(pipefds[1], data, size);
+        if (n < 0)
+                return -errno;
+        if ((size_t) n != size)
+                return -EIO;
+
+        (void) fd_nonblock(pipefds[0], false);
+
+        r = pipefds[0];
+        pipefds[0] = -1;
+
+        return r;
+
+try_dev_shm:
+        fd = open("/dev/shm", O_RDWR|O_TMPFILE|O_CLOEXEC, 0000);
+        if (fd < 0)
+                goto try_dev_shm_without_o_tmpfile;
+
+        n = write(fd, data, size);
+        if (n < 0)
+                return -errno;
+        if ((size_t) n != size)
+                return -EIO;
+
+        /* Let's reopen the thing, in order to get an O_RDONLY fd for the original O_RDWR one */
+        xsprintf(procfs_path, "/proc/self/fd/%i", fd);
+        r = open(procfs_path, O_RDONLY|O_CLOEXEC);
+        if (r < 0)
+                return -errno;
+
+        return r;
+
+try_dev_shm_without_o_tmpfile:
+        fd = mkostemp_safe(pattern);
```

Comment:
When I force falling back to this code, test-fd-util fails:
```sh
systemd-run --user -p SystemCallFilter='~memfd_create:ENOSYS' strace `pwd`/build/test-fd-util
```
<pre>Nov 16 15:13:59 strace[24961]: memfd_create(&quot;test&quot;, MFD_CLOEXEC)       = -1 ENOSYS (Function not implemented)
Nov 16 15:13:59 strace[24961]: openat(AT_FDCWD, &quot;/tmp&quot;, O_RDWR|O_EXCL|O_DIRECTORY|O_CLOEXEC|O_TMPFILE, 0600) = 3
Nov 16 15:13:59 strace[24961]: write(3, &quot;test\n&quot;, 5)                   = 5
Nov 16 15:13:59 strace[24961]: close(3)                                = 0
Nov 16 15:13:59 strace[24961]: memfd_create(&quot;data-fd&quot;, MFD_CLOEXEC|MFD_ALLOW_SEALING) = -1 ENOSYS (Function not implemented)
Nov 16 15:13:59 strace[24961]: openat(AT_FDCWD, &quot;/dev/shm&quot;, O_RDWR|O_DIRECTORY|O_CLOEXEC|O_TMPFILE, 000) = 3
Nov 16 15:13:59 strace[24961]: write(3, &quot;foo&quot;, 3)                      = 3
Nov 16 15:13:59 strace[24961]: openat(AT_FDCWD, &quot;/proc/self/fd/3&quot;, O_RDONLY|O_CLOEXEC) = -1 EACCES (Permission denied)
Nov 16 15:13:59 strace[24961]: close(3)                                = 0
Nov 16 15:13:59 strace[24961]: writev(2, [{iov_base=&quot;Assertion &apos;fd &gt;= 0&apos; failed at ..&quot;..., iov_len=104}, {iov_base=&quot;\n&quot;, iov_len=1}], 2Assertion &apos;fd &gt;= 0&apos; failed at ../src/test/test-fd-util.c:115, function test_acquire_data_fd(). Aborting.
Nov 16 15:13:59 strace[24961]: ) = 105
Nov 16 15:13:59 strace[24961]: rt_sigprocmask(SIG_UNBLOCK, [ABRT], NULL, 8) = 0
Nov 16 15:13:59 strace[24961]: rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1], [], 8) = 0
Nov 16 15:13:59 strace[24961]: getpid()                                = 24964
Nov 16 15:13:59 strace[24961]: gettid()                                = 24964
Nov 16 15:13:59 strace[24961]: tgkill(24964, 24964, SIGABRT)           = 0
Nov 16 15:13:59 strace[24961]: rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
Nov 16 15:13:59 strace[24961]: --- SIGABRT {si_signo=SIGABRT, si_code=SI_TKILL, si_pid=24964, si_uid=1000} ---
Nov 16 15:13:59 strace[24961]: +++ killed by SIGABRT (core dumped) +++
Nov 16 15:13:59 systemd-coredump[24982]: <font color="#CC0000"><b>Process 24964 (test-fd-util) of user 1000 dumped core.</b></font>
                                         
                                         <font color="#CC0000"><b>Stack trace of thread 24964:</b></font>
                                         <font color="#CC0000"><b>#0  0x00007f44a1b9c69b raise (libc.so.6)</b></font>
                                         <font color="#CC0000"><b>#1  0x00007f44a1b9e3b1 abort (libc.so.6)</b></font>
                                         <font color="#CC0000"><b>#2  0x00007f44a201b497 n/a (/home/zbyszek/src/systemd-work/build/src/shared/libsystemd-shared-235.so)</b></font>
                                         <font color="#CC0000"><b>#3  0x0000559a3a7d583a n/a (/home/zbyszek/src/systemd-work/build/test-fd-util)</b></font>
                                         <font color="#CC0000"><b>#4  0x0000559a3a7d5b6b n/a (/home/zbyszek/src/systemd-work/build/test-fd-util)</b></font>
                                         <font color="#CC0000"><b>#5  0x00007f44a1b8603a __libc_start_main (libc.so.6)</b></font>
                                         <font color="#CC0000"><b>#6  0x0000559a3a7d4caa n/a (/home/zbyszek/src/systemd-work/build/test-fd-util)</b></font>
Nov 16 15:13:59 systemd-coredump[25011]: <font color="#CC0000"><b>Process 24961 (strace) of user 1000 dumped core.</b></font>
</pre>

I'm not sure if this is somehow caused by the way I run the code or what.

Reply:
Yeah, `/proc/self/fd/3` is a symlink like `/dev/shm/#3865430 (deleted)`, so it cannot be re-opened.

Reply:
Maybe acquire_data_fd should be structured as a series of functions, so that it's actually possible to test the fallback code. Another option would be to install a seccomp filter to force the fallback code, but that's only easy with memfd_create. Either way, I think we need to have a way to automatically test the paths on different kernels.

Reply:
/proc/self/fd/3 is not a real symlink, it's magic. In my tests I had no problem to reopen deleted files like that. My own tests where done by adding a "goto" to the right label at the beginning of the fucntion and recompiling...

Reply:
That's what I did too.

Reply:
weird stuff. will look into it

Reply:
ah! it appears to be a perms thing! i.e. i specify the mode 0000 on the fd and unprivileged processes can't reopen it hence

Reply:
neat. a one byte fix... 0000 → 0500 and it works fine as unprivileged user too